cpp/src/fil/infer.cu

```diff
@@ -480,6 +487,115 @@ struct tree_aggregator_t<NITEMS, GROVE_PER_CLASS_MANY_CLASSES> {
+    return predict_proba ? phase1 + phase2 : std::max(phase1, phase2);
+  }
+  static size_t smem_accumulate_footprint(int num_classes) {
+    return sizeof(vec<NITEMS, float>) * num_classes *
+             max(1, FIL_TPB / num_classes) +
+           sizeof(vec<NITEMS, int>) * FIL_TPB + sizeof(int) * FIL_TPB;
+  }
+
+  __device__ __forceinline__ tree_aggregator_t(predict_params params,
+                                               void* accumulate_workspace,
+                                               void* finalize_workspace,
+                                               float* vector_leaf)
+    : num_classes(params.num_classes),
+      num_threads_per_class(max(1, blockDim.x / params.num_classes)),
+      vector_leaf(vector_leaf),
+      tmp_storage(finalize_workspace) {
+    // Assign workspace
+    char* ptr = (char*)accumulate_workspace;
+    per_class_margin = (vec<NITEMS, float>*)ptr;
+    ptr += sizeof(vec<NITEMS, float>) * num_classes * num_threads_per_class;
+    vector_leaf_indices = (vec<NITEMS, int>*)ptr;
+    ptr += sizeof(vec<NITEMS, int>) * blockDim.x;
+    thread_num_rows = (int*)ptr;
+
+    // Initialise shared memory
+    for (int i = threadIdx.x; i < num_classes * num_threads_per_class;
+         i += blockDim.x) {
+      per_class_margin[i] = vec<NITEMS, float>();
+    }
+    vector_leaf_indices[threadIdx.x] = vec<NITEMS, int>();
+    thread_num_rows[threadIdx.x] = 0;
+    // __syncthreads() is called in infer_k
+  }
+
+  __device__ __forceinline__ void accumulate(
+    vec<NITEMS, int> single_tree_prediction, int tree, int num_rows) {
+    // Perform a transpose in shared memory
+    // Assign each thread to a class, so they can accumulate without atomics
+    __syncthreads();
+    // Write indices to shared memory
+    vector_leaf_indices[threadIdx.x] = single_tree_prediction;
+    thread_num_rows[threadIdx.x] = num_rows;
+    __syncthreads();
+    // i here refers to each element of the matrix per_class_margin
+    for (int i = threadIdx.x; i < num_classes * num_threads_per_class;
+         i += blockDim.x) {
+      // if num_threads_per_class == 1, then c == i
+      int c = i % num_classes;
+      int row = i / num_classes;
+      // iterate over original thread inputs with stride num_threads_per_class
+      // j is the original thread input
+      // we have num_classes threads for each j
+      for (int j = row; j < blockDim.x; j += num_threads_per_class) {
+        for (int item = 0; item < thread_num_rows[j]; ++item) {
+          float pred =
+            vector_leaf[vector_leaf_indices[j][item] * num_classes + c];
+          per_class_margin[i][item] += pred;
+        }
+      }
+    }
+  }
+  __device__ __forceinline__ void finalize(float* out, int num_rows,
+                                           int num_outputs, output_t transform,
+                                           int num_trees,
+                                           int log2_threads_per_tree) {
+    __syncthreads();
```

Comment:
(see comment below first) and this __syncthreads() can be moved inside the `if`, since it's unnecessary when no data is moved out or into shared memory before the next __syncthreads()